### PR TITLE
[feat]: support certificate authentication and add tests

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
based on #648 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
support certificate authentication and add tests
* what changes does this pull request make?
support certificate authentication and add tests
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: crates/xline/src/storage/auth_store/store.rs [hunk 1]
```diff
@@ -1140,6 +1163,14 @@ where
     }
 }
 
+/// Get common name from tonic request
+fn get_cn<T>(request: &tonic::Request<T>) -> Option<String> {
+    let chain = request.peer_certs()?;
+    let cert_der = chain.first()?;
```
Comment (iGxnon) [inline crates/xline/src/storage/auth_store/store.rs:1169]: Is it possiable that a certificate chain has multiple certs here? Which one represents the proper CN?

File: crates/xline/tests/it/tls_test.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+use test_macros::abort_on_panic;
+use tonic::transport::{Certificate, ClientTlsConfig, Identity};
+use utils::config::{
+    AuthConfig, ClusterConfig, CompactConfig, LogConfig, MetricsConfig, StorageConfig, TlsConfig,
+    TraceConfig, XlineServerConfig,
+};
+use xline_client::types::kv::PutRequest;
+use xline_test_utils::{enable_auth, set_user, Cluster};
+
+#[tokio::test(flavor = "multi_thread")]
```
Comment (iGxnon) [inline crates/xline/tests/it/tls_test.rs:13]: When do we need to enable `multi_thread` for test?

File: crates/xline/src/server/xline_server.rs [hunk 3]
```diff
@@ -634,28 +635,58 @@ impl XlineServer {
...
                 let cert = fs::read_to_string(cert_path).await?;
                 let key = fs::read_to_string(key_path).await?;
                 Some(ServerTlsConfig::new().identity(Identity::from_pem(cert, key)))
             }
-            (None, None) => None,
-            _ => {
+            (_, Some(_), None) | (_, None, Some(_)) => {
                 return Err(anyhow!(
-                    "server_cert_path and server_key_path must be both set"
+                    "client_cert_path and client_key_path must be both set"
```
Comment (Phoenix500526) [inline crates/xline/src/server/xline_server.rs:686]: Should here be "server_cert_path and server_key_path must be both set"?

File: crates/utils/src/config.rs [hunk 4]
```diff
@@ -986,6 +986,13 @@ impl TlsConfig {
             client_key_path,
         }
     }
+
+    /// Whether the server tls is enabled
+    #[must_use]
+    #[inline]
+    pub fn server_tls_enabled(&self) -> bool {
+        self.server_cert_path.is_some() && self.server_key_path.is_some()
```
Comment (Phoenix500526) [inline crates/utils/src/config.rs:994]: "The server_cert seems a bit confusing, while the peer_cert is clearer. Just my two cents."

**Review by Phoenix500526 — COMMENTED** [review]
It's better to add an example in quick start to teach users how to use this feature.

File: crates/xline/src/storage/auth_store/store.rs (hunk 1 above)
Comment (themanforfree) [reply crates/xline/src/storage/auth_store/store.rs:1169]: The first one in the certificate chain is what we need

File: crates/xline/tests/it/tls_test.rs (hunk 2 above)
Comment (themanforfree) [reply crates/xline/tests/it/tls_test.rs:13]: Drop the `Cluster` need `multi_thread`

File: scripts/certgen.sh [hunk 5]
```diff
@@ -8,9 +8,17 @@ SERVER_KEY=${DIR}/certs/server.key
...
 SERVER_CRT=${DIR}/certs/server.crt
 
-CLIENT_KEY=${DIR}/certs/client.key
-CLIENT_CSR=${DIR}/certs/client.csr
-CLIENT_CRT=${DIR}/certs/client.crt
+ROOT_CLIENT_KEY=${DIR}/certs/root_client.key
+ROOT_CLIENT_CSR=${DIR}/certs/root_client.csr
+ROOT_CLIENT_CRT=${DIR}/certs/root_client.crt
+
+U1_CLIENT_KEY=${DIR}/certs/u1_client.key
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply scripts/certgen.sh:20]: Added

File: crates/xline/src/server/xline_server.rs (hunk 3 above)
Comment (themanforfree) [reply crates/xline/src/server/xline_server.rs]: Fixed

File: crates/utils/src/config.rs (hunk 4 above)
Comment (themanforfree) [reply crates/utils/src/config.rs]: Renamed

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by iGxnon — APPROVED** [review]

**Pull request merged**